plutus-core/plutus-core/src/PlutusCore/Builtin/Debug.hs

```diff
@@ -21,31 +23,60 @@ import PlutusCore.Name as Export
 --
 -- >>> :t elaborateDebug False
 -- elaborateDebug False :: Bool
--- >>> :t elaborateDebug $ \_ -> ()
--- elaborateDebug $ \_ -> ()
+-- >>> :t elaborateDebug fst
+-- elaborateDebug fst
+--   :: (TyVarRep ('TyNameRep "a" 0), TyVarRep ('TyNameRep "b" 1))
+--      -> TyVarRep ('TyNameRep "a" 0)
+elaborateDebug
+    :: forall a j. ElaborateFromTo 0 j (Term TyName Name DefaultUni DefaultFun ()) a
+    => a -> a
+elaborateDebug = id
+
+-- >>> :t makeBuiltinMeaningDebug $ \_ -> ()
+-- makeBuiltinMeaningDebug $ \_ -> ()
 --   :: Opaque
 --        (Term TyName Name DefaultUni DefaultFun ())
 --        (TyVarRep ('TyNameRep "a" 0))
 --      -> ()
--- >>> :t elaborateDebug $ Just ()
+-- >>> :t makeBuiltinMeaningDebug False
+-- <interactive>:1:1: error:
+--     • A built-in function must take at least one type or term argument
+--       ‘Bool’ is a built-in type so you can embed any of its values as a constant
+--       If you still want a built-in function, add a dummy ‘()’ argument
+--     • In the expression: makeBuiltinMeaningDebug False
+-- >>> :t makeBuiltinMeaningDebug $ 0 + 42
+-- <interactive>:1:27: error:
+--     • Built-in functions are not allowed to have constraints
+--       To fix this error instantiate all constrained type variables
+--     • In the second argument of ‘($)’, namely ‘0 + 42’
+--       In the expression: makeBuiltinMeaningDebug $ 0 + 42
+-- >>> :t makeBuiltinMeaningDebug $ Just ()
 -- <interactive>:1:1: error:
 --     • No instance for ‘KnownTypeAst DefaultUni (Maybe ())’
 --       Which means
 --         ‘Maybe ()’
 --       is neither a built-in type, nor one of the control types.
 --       If it can be represented in terms of one of the built-in types
---         then go add the instance (you may need a ‘KnownTypeIn’ one too)
+--         then go add the instance (you may need a few others too)
 --         alongside the instance for the built-in type.
 --       Otherwise you may need to add a new built-in type
 --         (provided you're doing something that can be supported in principle)
---     • In the expression: elaborateDebug $ Just ()
--- >>> :t elaborateDebug $ 0 + 42
--- <interactive>:1:18: error:
---     • Built-in functions are not allowed to have constraints
---       To fix this error instantiate all constrained type variables
---     • In the second argument of ‘($)’, namely ‘0 + 42’
---       In the expression: elaborateDebug $ 0 + 42
-elaborateDebug
-    :: forall a j. ElaborateFromTo 0 j (Term TyName Name DefaultUni DefaultFun ()) a
+--     • In the expression: makeBuiltinMeaningDebug $ Just ()
+-- >>> :t makeBuiltinMeaningDebug fst
+-- <interactive>:1:1: error:
+--     • Plutus type variables can't directly appear inside built-in types
```

Comment:
This is confusing, because it is possible to call `makeBuiltinMeaning $ \a b -> a`, but not `makeBuiltinMeaning $ \(a, b) -> a`. Could you rephrase it to explain the difference between the two?

Reply:
It's a type error, it's not meant to teach the user how and why things are aligned, we have proper documentation on that in the `PlutusCore.Default.Builtin` module. It is indeed unfortunate that we support `\a b -> a` and don't support `\(a, b) -> a`, but there are good reasons for that and I can't fit them into a type error, we have loads of docs with such explanations.

Even distinguishing between `a -> b -> a` and `(a, b) -> a` would require me to define what a built-in type is, talk about how `(Integer, Bool)` is different from `(a, b)` (since `(Integer, Bool) -> Integer` is perfectly fine) etc. It's just not feasible within a type error.

Think of our custom type errors as reminders. They are meant to remind you of something that you're already supposed to be familiar with and also give you quick advice on how to proceed.

Or do you maybe have a suggestion on how to improve the message?

Reply:
Fair enough. I don't actually know why one is allowed and the other isn't (I should revisit the docs) so I don't have a suggestion. But I was wondering if there's a good one sentence summary like "Can't call `makeBuiltinMeaning` on `\(a, b) -> a` because ...", like GHC's "because type variable `a` is untouchable". It doesn't explain what "untouchable" means (there's a paper for that) but one can then go and search for it.

But adding a builtin function is not something we frequently do so this is all fine.

Reply:
> "Can't call `makeBuiltinMeaning` on `\(a, b) -> a` because ...", like GHC's "because type variable `a` is untouchable".

Problem is, we piggyback on existing machinery and this is where we throw this error:

```haskell
instance TypeError NoStandalonePolymorphicDataErrMsg => uni `Contains` TyVarRep where
    knownUni = underTypeError
```

We don't have any information as to what exactly went wrong, we only know that this path is accessed when some built-in type got applied to a type variable, so we can't be specific. Maybe we can find another way of throwing the error, though.

I think "type variables can't appear inside built-in types" is a good one sentence summary, but I like the untouchable idea. But then should I add such a keyword to every other error message?

Gonna summon @michaelpj.

Reply:
Anyways. this seems like a separate concern, so I'm gonna press "merge".

Reply:
Ah so "can't appear inside built-in types" in this case means "can't appear inside `(,)`". Ok, that makes sense, but I think a more explicit way of saying it would be "can't be an argument of a built-in type constructor", or "must only be used as the entire type of an argument".

Reply:
Great suggestion, thank you. [Tweaked](https://github.com/input-output-hk/plutus/blob/177e942e4650314b52665e288bbe7538305c4a8b/plutus-core/plutus-core/src/PlutusCore/Builtin/Polymorphism.hs#L200-L205) the wording.